File: Programmers/Lv2/전력망을 둘로 나누기.py
from collections import deque
def solution(n, wires):
    answer = int(1e9)
    
    graph = [[] for _ in range(n + 1)]
    for wire in wires:
        graph[wire[0]].append(wire[1])
        graph[wire[1]].append(wire[0])
    
    def bfs(s, endPoint):
        visited = [0] * (n+1)
        visited[s] = 1
        q = deque()
        q.append(s)
        
        cnt = 1
        
        while q:
            x = q.popleft()
            
            for node in graph[x]:
                if node != endPoint:
                    if visited[node] == 0:
                        q.append(node)
                        visited[node] = 1
                        cnt += 1
        return cnt
                    
    
    for v in range(1, n + 1):
        for node in graph[v]:
            a = bfs(v, node)
            b = bfs(node, v)
            answer = min(answer, abs(a-b))
    
    return answer

File: Programmers/Lv2/전력망을 둘로 나누기.js
function solution(n, wires) {
    var answer = 1e9;
    
    var graph = Array.from(Array(n + 1), () => Array());
    for(wire of wires){
        let a = wire[0];
        let b = wire[1];
        
        graph[a].push(b)
        graph[b].push(a)
    }
    
    const bfs = (start, cant) => {
        var q = [start];
        var visited = new Set();
        visited.add(start)
        
        var cnt = 1;
        
        while(q.length > 0){
            var node = q.shift();
            
            graph[node].forEach(v => {
                if(v !== cant){
                    if(!visited.has(v)){
                        q.push(v);
                        visited.add(v)
                        cnt++;
                    }
                }
                
            })
        }
        
        return cnt;
    }
    
    for(let v = 1; v < n + 1; v++){
        graph[v].forEach(e => {
            var a = bfs(e, v);
            var b = bfs(v, e);
            answer = Math.min(answer, Math.abs(a-b));
        })
    }
    
    
    return answer;
}